[CH82] Support Internationalization

seerocode opened the pull request:
## [Clubhouse link](https://app.clubhouse.io/stanford-computational-policy-lab/story/82/support-internationalization)

### Description

- Added library [React i18n](https://github.com/i18next/react-i18next) library for internationalization support that includes browser language detector and lazy loading translations 
- Added locale for en-GB to support UK English and custom translations

Commit:
Add react18n packages [ch82]

Commit:
Add i18n service [ch82]
- Add en-GB locale for UK English support
- Add sample translation for program

seerocode commented:
Missing tests

Commit:
Update App and sidebar for internationalization [ch82]
- Add stand-in suspense fallback to App to lazy load translations
- Rename AppSidebar teams to program to test locale translation

jnu commented on client/src/App.tsx:
```diff
@@ -1,4 +1,4 @@
-import React from "react";
+import React, { Suspense } from "react";
```
😲 

very nice! this is one of my fav new features of react!

seerocode replied:
I didn't know what this was until an error told me to add it in. The loading element is a placeholder though and we should add a real one in.

jnu commented on client/src/components/AppSidebar.tsx:
```diff
@@ -38,3 +45,5 @@ export default class AppSidebar extends React.Component {
   }
 }
+
+export default withTranslation("en-gb")(AppSidebar);
```
for my info, what is the `en-gb` doing here? and how does this relate to the `i18next.ts` below where we're setting up to detect locale from the browser?

seerocode replied:
This is supposed to specify the use of the en-gb translations we have in the locales folder. So if the local detected is en-gb, it will use these translations. But tests are necessary to make sure it works because UI-wise, it's not. :/

seerocode replied:
<strike>Oh, I think I was using the wrong thing! Should probably be `withNamespaces()` instead: https://react.i18next.com/legacy-v9/step-by-step-guide</strike>

NVM!

jnu commented on client/src/i18n/i18next.ts:
```diff
@@ -0,0 +1,32 @@
+  // for all options read: https://www.i18next.com/overview/configuration-options
+  .init({
+    lng: "en",
+    fallbackLng: "en",
```
should this be `en-gb` since that's the only one we've defined so far?

seerocode replied:
I can change it!

jnu replied:
this looks fine to me now -- were you able to figure out how to change the language? was it some cookie issue?

seerocode replied:
@jnu Just pushed a change to correct it! The `lng` key in the i18next setup overrides language detection so I removed it.

jnu reviewed (approved):
looks good!!

I've never used this library before tho and i just want to confirm that we are definitely auto-detecting locale from the browser, and loading the correct file (if available).

Commit:
Fix jest dependency version

Commit:
Convert AppSidebar to functional component

Commit:
Reconfigure i18n
- Move translation files to public folder and rename
- Add translation for english
- Remove backend options from i18n config

Commit:
Add i18n test for AppSidebar

Commit:
Update AppSidebar test

seerocode commented:
Tests added!

jnu commented on client/src/__tests__/AppSidebar.test.tsx:
```diff
@@ -0,0 +1,40 @@
+
+  test('translates sidebar title text to UK English', () => {
+    const sidebar = shallow(<AppSidebar />);
+    expect(sidebar.find(SubMenu).first().prop('title')).toBe('My Programmes');
```
🤗 😍 yay tests

jnu reviewed (approved):
looks great to me!

Commit:
Remove lng key to prevent language detection override
- Tested in Google Chrome incognito by changing default language
